rapids-cmake/test/generate_resource_spec.cmake

```diff
@@ -0,0 +1,94 @@
+#=============================================================================
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+rapids_test_generate_resource_spec
+----------------------------------
+
+.. versionadded:: v23.04.00
+
+Generates a JSON resource specification file representing the machine's GPUs
+using system introspection.
+
+  .. code-block:: cmake
+
+    rapids_test_generate_resource_spec( DESTINATION filepath )
+
+Generates a JSON resource specification file representing the machine's GPUs
+using system introspection. This will allow CTest to schedule multiple
+single-GPU tests in parallel on multi-GPU machines.
+
+For the majority of projects :cmake:command:`rapids_test_init` should be used.
+This command should be used directly projects that require multiple spec
+files to be generated.
+
+``DESTINATION``
+  Location that the JSON output from the detection should be written to
+
+.. note::
+    Unlike rapids_test_init this doesn't set CTEST_RESOURCE_SPEC_FILE
+
+#]=======================================================================]
+function(rapids_test_generate_resource_spec DESTINATION filepath)
```

Comment:
No we need to regenerate each time we invoke CMake since the user can execute CMake with `CUDA_VISIBLE_DEVICES` set to narrow the set of GPUs they want used.

Reply:
I didn't consider that option, that is a nice use case to support.